### PR TITLE
Attempt to add upgradeConfig support

### PR DESCRIPTION
**Pull request by darioush**
This will help with testing upgradeBytes and possibly using this in `avalanche-cli`

### CODE REVIEW

File: cmd/control/control.go [hunk 1]
```diff
@@ -507,6 +508,12 @@ func newAddNodeCommand() *cobra.Command {
 		"",
 		"[optional] JSON string of map that maps from chain id to its config file contents",
 	)
+	cmd.PersistentFlags().StringVar(
+		&upgradeConfigs,
+		"upgrade-configs",
+		"",
+		"[optional] JSON string of map that maps from chain id to its upgrade file contents",
```
Comment (holisticode) [inline cmd/control/control.go:515]: Nit: "JSON string of map from chain id to..."

File: cmd/control/control.go [hunk 2]
```diff
@@ -587,6 +602,18 @@ func newRestartNodeCommand() *cobra.Command {
 		"",
 		"whitelisted subnets (comma-separated)",
 	)
+	cmd.PersistentFlags().StringVar(
+		&chainConfigs,
+		"chain-configs",
+		"",
+		"[optional] JSON string of map that maps from chain id to its config file contents",
```
Comment (holisticode) [inline cmd/control/control.go:609]: same

File: cmd/control/control.go [hunk 3]
```diff
@@ -587,6 +602,18 @@ func newRestartNodeCommand() *cobra.Command {
...
+	)
+	cmd.PersistentFlags().StringVar(
+		&upgradeConfigs,
+		"upgrade-configs",
+		"",
+		"[optional] JSON string of map that maps from chain id to its upgrade file contents",
```
Comment (holisticode) [inline cmd/control/control.go:615]: ditto

File: cmd/control/control.go (hunk 2 above)
Comment (holisticode) [reply cmd/control/control.go:609]: I thought chain configs were already supported, @felipemadero can you confirm?

File: local/network.go [hunk 4]
```diff
@@ -1099,17 +1109,23 @@ func (ln *localNetwork) buildFlags(
 		return buildFlagsReturn{}, err
 	}
 
-	// Use random free API port unless given in config file
-	apiPort, err := getPort(nodeConfig.Flags, configFile, config.HTTPPortKey)
-	if err != nil {
-		return buildFlagsReturn{}, err
+	// Prefer API port and P2P (staking) port from nodeConfig
+	apiPort, p2pPort := nodeConfig.ApiPort, nodeConfig.StakingPort
```
Comment (holisticode) [inline local/network.go:1113]: What if those ports were already taken?

**Review by holisticode — COMMENTED** [review]
To me it looks basically good, I have only a couple of nits, thanks for this addition.

I'd like @felipemadero to check on the chain config changes, as I thought they were already supported.

**Review by felipemadero — COMMENTED** [review]
Could you update the PR to latest changes? Will do a review tomorrow

File: local/network.go (hunk 4 above)
Comment (darioush) [reply local/network.go:852]: The ports would be used by the same node previously, so it seems reasonable to assume they are available after the node shuts down?
Not sure if you already handled this is a different way but it is useful eg to not have to reconfigure metamask or other clients to submit tx after restarting the node.

Comment (felipemadero) [reply local/network.go]: Yeah it is already handled in other way in ANR

File: cmd/control/control.go [hunk 5]
```diff
@@ -163,6 +164,12 @@ func newStartCommand() *cobra.Command {
 		"",
 		"[optional] JSON string of map from chain id to its config file contents",
 	)
+	cmd.PersistentFlags().StringVar(
+		&upgradeConfigs,
+		"upgrade-configs",
+		"",
+		"[optional] JSON string of map from chain id to its config file contents",
```
Comment (felipemadero) [inline cmd/control/control.go:171]: nit: upgrade file contents

**Pull request merged**